Automattic/wp-calypso
# Force the iframe to redirect if it fails to reload

#### Changes proposed in this Pull Request

This is an alternative approach to https://github.com/Automattic/wp-calypso/pull/43127.

In Safari, if you have "Prevent cross-site tracking" enabled, the iframe containing the block editor fails to load - it gets stuck in an infinite loop. Unfortunately this means that the `onLoad` event of the iframe never fires which means that the user is stuck on the placeholder screen.

To fix this I added a `setInterval` which listens for when the `iframe` is ready to load. Then we set a timeout for six seconds (one second longer than the timeout on the `iframe`). If the iframe loads then we cancel the timeout, otherwise we redirect the user to the iframe address.

The advantage of this approach over #43127 is that it ensures that users will never get stuck in the iframed editor, if something else goes wrong, or if other browsers start to block third party cookies.

#### Testing instructions
I haven't found a clean way to test this.

In order to test in Safari with "Prevent cross-site tracking" enabled you have to run Calypso locally on the wordpress.com domain. There are instructions on how to do this on the field guide (I ended up using Charles). The problem is that the block editor always tries to do a redirect to wordpress.com/wp-login.php; if this is pointing at Calypso then it throws a 404. In Safari a 404 does trigger the `onLoad` event for the iframe, so with this setup you won't see the problem.

After the redirect to `wordpress.com/wp-login.php` you are redirected back to `https://wordpress.com/block-editor/post/sitebuilder197760975.wpcomstaging.com?authWpAdmin=true`. If you visit this URL directly then you can skip this step, and load the iframe. However the iframe will still load `wordpress.com/wp-login.php` inside it, which will again throw a 404 and trigger the `onLoad` event.

However if you can replicate an infinite loop locally you can update the iframe URL to point to your local infinite loop, which will then trigger the iframe error state.

I used this: ```<?php header("Location: http://localhost:4444/redirect.php"); ?>```

Of course if you do this you will be redirected to the page with the infinite loop, which will throw an error :)

Given the complexities in testing this, I wonder if it makes more sense to test it in staging...

Fixes #43068
